File: task3_gpt.cpp
#include <iostream>
#include <vector>
#include <stdexcept>
#include <chrono>
#include <sstream>
#include <numeric>
using namespace std;

vector<vector<int>> parseMatrixInput(const string& input, int& rows, int& cols) {
    vector<vector<int>> matrix;
    stringstream ss(input);
    string rowStr;
    while (getline(ss, rowStr, ';')) {
        vector<int> row;
        stringstream rowStream(rowStr);
        int value;
        while (rowStream >> value) {
            row.push_back(value);
        }
        if (matrix.empty()) {
            cols = row.size();
        }
        matrix.push_back(row);
        rows++;
    }
    return matrix;
}

vector<vector<int>> matrixMultiplication(const vector<vector<int>>& A, const vector<vector<int>>& B) {
    if (A[0].size() != B.size()) {
        throw invalid_argument("Number of columns in A must match number of rows in B.");
    }

    vector<vector<int>> result(A.size(), vector<int>(B[0].size(), 0));
    for (int i = 0; i < A.size(); i++) {
        for (int j = 0; j < B[0].size(); j++) {
            for (int k = 0; k < A[0].size(); k++) {
                result[i][j] += A[i][k] * B[k][j];
            }
        }
    }
    return result;
}

int main() {
    string inputA, inputB;
    int rowsA = 0, colsA = 0, rowsB = 0, colsB = 0;

    cout << "Enter the matrix A rows separated by semicolons (e.g. '1 2; 3 4'): ";
    getline(cin, inputA);
    vector<vector<int>> A = parseMatrixInput(inputA, rowsA, colsA);

    cout << "Enter the matrix B rows separated by semicolons (e.g. '1 2; 3 4'):  ";
    getline(cin, inputB);
    vector<vector<int>> B = parseMatrixInput(inputB, rowsB, colsB);

    auto start = chrono::high_resolution_clock::now();
    vector<vector<int>> result = matrixMultiplication(A, B);
    auto end = chrono::high_resolution_clock::now();

    cout << "The result of matrix multiplication is:" << endl;
    for (const auto& row : result) {
        for (int element : row) {
            cout << element << " ";
        }
        cout << endl;
    }

    chrono::duration<double, std::milli> duration = end - start;
    cout << "Time spent on Execution: " << duration.count() << " ms" << endl;

    return 0;
}

File: task3_gpt.py
import time

def get_matrix_input():
 
    raw_input = input("Enter the matrix rows separated by semicolons (e.g. '1 2; 3 4'): ")
    matrix = [list(map(int, row.split())) for row in raw_input.split(';')]
    return matrix

def matrix_multiplication(A, B):

    rows_A, cols_A = len(A), len(A[0])
    rows_B, cols_B = len(B), len(B[0])

    if cols_A != rows_B:
        raise ValueError("Number of columns in A must match number of rows in B.")

    result = [[sum(A[i][k] * B[k][j] for k in range(cols_A)) for j in range(cols_B)] for i in range(rows_A)]
    return result

def main():
   
    print("Input for Matrix A:")
    A = get_matrix_input()
    print("Input for Matrix B:")
    B = get_matrix_input()

    start_time = time.time()
    result = matrix_multiplication(A, B)
    end_time = time.time()
    
    print("The result of matrix multiplication is:")
    for row in result:
        print(row)
        
        
    execution_time = end_time - start_time
    print(f"Time spent on Execution: {execution_time:.6f} seconds")

if __name__ == "__main__":
    main()
